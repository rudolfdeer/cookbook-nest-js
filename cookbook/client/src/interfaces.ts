export interface IState {
  recipes: IRecipe[];
  cookbooks: ICookbook[];
  user: IUser | null;
}

export interface IUser {
  id: number;
  name: string;
  email: string;
  password: string;
  image: string;
  bio: string;
  Recipe_Saveds?: IRecipeSaved[];
  Cookbook_Saveds?: ICookbookSaved[];
}

export interface IRecipe {
  id: number;
  title: string;
  description: string;
  image: string;
  directions: string[];
  ingredients: string[];
  time: number;
  views: number;
  UserId: number;
  User: IUser;
  Recipe_Comments: IRecipeComment[];
  Recipe_Likes: IRecipeLike[];
}

export interface ICookbook {
  id: number;
  title: string;
  description: string;
  image: string;
  tags: string[];
  views: number;
  UserId: number;
  User: IUser;
  Recipe_Cookbooks: IRecipeCookbook[];
  Cookbook_Comments: ICookbookComment[];
  Cookbook_Likes: ICookbookLike[];
}

export interface IRecipeLike {
  RecipeId: number;
  UserId: number;
}

export interface IRecipeComment {
  id: number;
  text: string;
  date: string;
  RecipeId: number;
  UserId: number;
  User?: IUser;
}

export interface IRecipeSaved {
  RecipeId: number;
  UserId: number;
  Recipe?: IRecipe;
}

export interface IRecipeCookbook {
  RecipeId: number;
  CookbookId: number;
  Recipe: IRecipe;
}

export interface ICookbookLike {
  CookbookId: number;
  UserId: number;
}

export interface ICookbookSaved {
  CookbookId: number;
  UserId: number;
  Cookbook?: ICookbook;
}

export interface ICookbookComment {
  id: number;
  text: string;
  date: string;
  CookbookId: number;
  UserId: number;
  User?: IUser;
}

export interface IRecipeRequestBody {
  title: string;
  description: string;
  ingredients: string;
  directions: string;
  time?: number;
  views?:number;
  likeUserIds?: number[];
}

export interface ICookbookRequestBody {
  title: string;
  description: string;
  recipesIds: number[];
  tags?: string[];
  views?: number;
  likeUserIds?: number[];
  image?: File;
}

export interface IUserRequestBody {
  name?: string;
  photo?: string;
  bio?: string;
  savedRecipesIds?: number[];
  savedCookbooksIds?: number[];
}

export interface IAuthRequestBody {
  email: string;
  password: string;
}

export interface ISearchListItem {
  id: number;
  name: string;
}

export interface ISignUpForm {
  email: string;
  password: string;
  confirm: string;
}
