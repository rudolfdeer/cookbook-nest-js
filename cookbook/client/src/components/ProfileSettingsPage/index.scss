@import '../../styles/index.scss';

.profile-page {
  &--settings {
    padding: 104px 0;
    background-color: $grey-light;

    &__content {
      margin-bottom: 18px;
      display: flex;
      flex-wrap: wrap;
      width: 100%;
    }

    &__photo {
      &--editable {
        margin-right: 40px;
        width: 160px;
        height: 160px;
        overflow: hidden;
      }

      &__label {
        
        position: relative;
        display: flex;
        margin-right: 40px;
        width: 100%;
        height: 100%;
        border-radius: 102px;
        background-color: #181818;
        background: url('../../assets/images/photo-mask.png') center no-repeat;
        z-index: 221;
      }

      &__input {
        opacity: 0;
        appearance: none;
        display: block;
        height: 100%;
        width: 100%;
        position: absolute;
        z-index: 222;
        border-radius: 102px;

        &:hover {
          cursor: pointer;
        }

      }

      &__image {
        &--opacity {
          display: block;
          margin: auto;
          width: 100%;
          height: 100%;
          object-fit: cover;
          z-index: -1;
          border-radius: 102px;
          opacity: 0.5;
        }
      }
    }

    &__user {
      display: flex;
      flex-direction: column;
      width: 100%;
      max-width: 580px;

      &__name {
        @include title-40;
      }

      &__bio {
        &--editable {
          @include text-18;
          min-height: 180px;
          @include text-18;

          width: 100%;
          margin-right: 40px;
          border: none;
          resize: none;
          display: flex;
          min-height: 180px;
          padding: 0;

          &:disabled {
            background: none;
          }
        }
      }

      &__form {
        display: flex;
        justify-content: space-between;
        width: 100%;
        height: 100%;

        &__input {
          border: none;

          &:disabled {
            background: none;
          }

          &--submit {
            @include text-18;
            height: 22px;

            background: none;
            border: none;
            color: $primary-yellow;

            &:hover {
              cursor: pointer;
              text-decoration: underline;
            }
          }
        }
      }
    }

    &__nav {
      margin-bottom: 40px;
      height: 62px;
      display: flex;
      align-items: flex-end;

      .nav__list {
        @include page-nav-list;

        .list__item {
          &--selected {
            margin-right: 0;
          }
        }
      }
    }

    &__info {
      padding: 72px 56px;
      background-color: $white;

      &__title {
        @include title-40;

        margin-bottom: 32px;
      }

      &__form {
        margin-bottom: 24px;

        &:last-child {
          margin-bottom: 0;
        }

        &__label {
          @include text-18;
          font-weight: bold;
          display: inline-block;
          margin-right: 72px;
          width: 150px;
        }

        &__input {
          @include text-18;
          margin-right: 16px;

          border: none;
          border-bottom: $main 1px solid;
          background: $grey-light;

          &:disabled {
            background: none;
          }
        }
      }
    }

    &__btns {
      display: flex;
      width: 100%;

      &__btn {
        &--logout {
          @include button-dark;

          margin-right: 40px;

          height: 32px;
          //width: 120px;
        }

        &--delete {
          @include text-18;
          height: 32px;
          border: none;
          background: none;
          color: red;

          &:hover {
            cursor: pointer;
            text-decoration: underline;
          }
        }
      }
    }
  }
}

@media (max-width: 1199px) {
  .profile-page {
    &--settings {
      &__content {
        flex-direction: column;
        align-items: center;
      }

      &__photo {
        &--editable {
          margin-right: 0;
          margin-bottom: 24px;
        }
      }

      &__user {
        align-items: center;
      }

      &__info {
        &__form {
          &__label {
            margin-right: 24px;
          }
        }
      }

      &__nav  {
        justify-content: center;
      }
    }
  }
}

@media (max-width: 639px) {
  .profile-page {
    &--settings {
      &__content {
        justify-content: center;
      }

      &__nav {
        display: block;
        width: 100%;
      }

      &__photo {
        &--editable {
          margin-right: 0;
          margin-bottom: 24px;
        }
      }

      &__user {
        align-items: center;

        &__form {
          flex-direction: column;
          align-items: center;
          margin-bottom: 24px;
        }

        &__bio {
          &--editable {
            margin-right: 0;
            margin-bottom: 10px;
          }
        }
      }

      &__info {
        padding: 24px;
        &__form {
          display: flex;
          flex-direction: column;
          align-items: center;

          &__label {
            width: auto;
            margin-right: 0;
            
          }

          &__input {
            margin-right: 0;
            text-align: center;
          }
        }
      }

      &__btns {
        flex-direction: column;
        align-items: center;

        &__btn {
          &--logout {
            margin-right: 0;
            margin-bottom: 10px;
          }
        }
      }
    }
  }
}