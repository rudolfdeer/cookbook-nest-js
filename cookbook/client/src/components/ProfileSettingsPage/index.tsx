import React, { useState } from 'react';
import { Link, Redirect } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import ROUTES from '../../constants/routes';
import Footer from '../Footer';
import HeaderConnect from '../../redux/containers/HeaderConnect';
import { IUser, IUserRequestBody } from '../../interfaces';
import SERVER_URL from '../../constants/serverUrl';

import './index.scss';

type ProfileSettingsPageProps = {
  user: IUser;
  deleteUser: () => Promise<void>;
  updateUser: (data: IUserRequestBody) => Promise<void>;
  changePassword: (password: string) => Promise<void>;
  changeEmail: (email: string) => Promise<void>;
  updateUsersPhoto: (data: FormData) => Promise<void>;
  signOut: () => Promise<void>;
};

export default function ProfileSettingsPage(
  props: ProfileSettingsPageProps,
): JSX.Element {
  if (!props.user) {
    return <Redirect to={ROUTES.HOME} />;
  }

  const { t } = useTranslation();

  const {
    user,
    deleteUser,
    updateUser,
    changePassword,
    changeEmail,
    signOut,
    updateUsersPhoto,
  } = props;
  const {
    name, email, bio, image,
  } = user;

  const [isBioDisabled, setBioDisabled] = useState(true);
  const [isNameDisabled, setNameDisabled] = useState(true);
  const [isEmailDisabled, setEmailDisabled] = useState(true);
  const [isPasswordDisabled, setPasswordDisabled] = useState(true);
  const [newBio, setNewBio] = useState(bio);
  const [newName, setNewName] = useState(name);
  const [newEmail, setNewEmail] = useState(email);
  const [newPassword, setNewPassword] = useState('');
  const [photoSrc, setPhotoSrc] = useState(
    `${SERVER_URL}/${image}` || './assets/images/photo-mask.png',
  );

  const onPhotoChange = async (e: React.FormEvent) => {
    const input = e.target as HTMLInputElement;
    const file = input.files[0];

    const data = new FormData();
    data.append('photo', file);
    await updateUsersPhoto(data);

    const reader = new FileReader();
    reader.onload = () => {
      const result = String(reader.result);
      setPhotoSrc(result);
    };

    reader.readAsDataURL(file);
  };

  return (
    <>
      <div className="wrapper">
        <HeaderConnect />
      </div>
      <main className="profile-page--settings">
        <div className="wrapper">
          <section className="profile-page--settings__content">
            <div className="profile-page--settings__photo--editable">
              <label
                htmlFor="avatar"
                className="profile-page--settings__photo__label"
              >
                <input
                  type="file"
                  className="profile-page--settings__photo__input"
                  name="photo"
                  onChange={(e) => onPhotoChange(e)}
                />
                <img
                  src={photoSrc}
                  alt="User photo"
                  className="profile-page--settings__photo__image--opacity"
                />
              </label>
            </div>
            <div className="profile-page--settings__user editable">
              <div className="profile-page--settings__user__name">
                {newName}
              </div>
              <form action="" className="profile-page--settings__user__form">
                <textarea
                  name="bio"
                  value={newBio}
                  className="profile-page--settings__user__bio--editable"
                  disabled={isBioDisabled}
                  onChange={(e) => {
                    const target = e.target as HTMLTextAreaElement;
                    setNewBio(target.value);
                  }}
                />
                {isBioDisabled ? (
                  <button
                    className="profile-page--settings__user__form__input--submit"
                    onClick={(e) => {
                      e.preventDefault();
                      setBioDisabled(false);
                    }}
                  >
                    {t('EDIT_BTN')}
                  </button>
                ) : (
                  <input
                    type="submit"
                    className="profile-page--settings__user__form__input--submit"
                    value={t('SAVE_BTN') as string}
                    onClick={(e) => {
                      e.preventDefault();
                      setBioDisabled(true);
                      updateUser({
                        name: newName,
                        bio: newBio,
                      });
                    }}
                  />
                )}
              </form>
            </div>
          </section>
          <nav className="profile-page--settings__nav">
            <ul className="nav__list">
              <li className="list__item">
                <Link to={ROUTES.PROFILE_SAVED}>{t('SAVED')}</Link>
              </li>
              <li className="list__item">
                <Link to={ROUTES.PROFILE_COOKBOOKS}>{t('MY_COOKBOOKS')}</Link>
              </li>
              <li className="list__item">
                <Link to={ROUTES.PROFILE_RECIPES}>{t('MY_RECIPES')}</Link>
              </li>
              <li className="list__item--selected">{t('MY_SETTINGS')}</li>
            </ul>
          </nav>
          <section className="profile-page--settings__info">
            <div className="profile-page--settings__info__title">
              {t('PERSONAL_INFORMATION')}
            </div>
            <form action="" className="profile-page--settings__info__form">
              <label
                htmlFor="name"
                className="profile-page--settings__info__form__label"
              >
                {t('NAME')}
              </label>
              <input
                type="text"
                name="name"
                value={newName}
                className="profile-page--settings__info__form__input"
                disabled={isNameDisabled}
                onChange={(e) => {
                  const target = e.target as HTMLInputElement;
                  setNewName(target.value);
                }}
              />
              {isNameDisabled ? (
                <input
                  type="submit"
                  className="profile-page--settings__user__form__input--submit"
                  value={t('EDIT_BTN') as string}
                  onClick={(e) => {
                    e.preventDefault();
                    setNameDisabled(false);
                  }}
                />
              ) : (
                <input
                  type="submit"
                  className="profile-page--settings__user__form__input--submit"
                  value={t('SAVE_BTN') as string}
                  onClick={(e) => {
                    e.preventDefault();
                    setNameDisabled(true);
                    updateUser({
                      name: newName,
                      photo: photoSrc,
                      bio: newBio,
                    });
                  }}
                />
              )}
            </form>
            <form action="" className="profile-page--settings__info__form">
              <label
                htmlFor="email"
                className="profile-page--settings__info__form__label"
              >
                {t('EMAIL')}
              </label>
              <input
                type="text"
                name="email"
                value={newEmail}
                className="profile-page--settings__info__form__input"
                disabled={isEmailDisabled}
                onChange={(e) => {
                  const target = e.target as HTMLInputElement;
                  setNewEmail(target.value);
                }}
              />
              {isEmailDisabled ? (
                <input
                  type="submit"
                  className="profile-page--settings__user__form__input--submit"
                  value={t('EDIT_BTN') as string}
                  onClick={(e) => {
                    e.preventDefault();
                    setEmailDisabled(false);
                  }}
                />
              ) : (
                <input
                  type="submit"
                  className="profile-page--settings__user__form__input--submit"
                  value={t('SAVE_BTN') as string}
                  onClick={(e) => {
                    e.preventDefault();
                    setEmailDisabled(true);
                    changeEmail(newEmail);
                  }}
                />
              )}
            </form>
            <form action="" className="profile-page--settings__info__form">
              <label
                htmlFor="password"
                className="profile-page--settings__info__form__label"
              >
                {t('PASSWORD')}
              </label>
              <input
                type="password"
                name="password"
                value={newPassword}
                className="profile-page--settings__info__form__input"
                disabled={isPasswordDisabled}
                onChange={(e) => {
                  const target = e.target as HTMLInputElement;
                  setNewPassword(target.value);
                }}
              />
              {isPasswordDisabled ? (
                <input
                  type="submit"
                  className="profile-page--settings__user__form__input--submit"
                  value={t('EDIT_BTN') as string}
                  onClick={(e) => {
                    e.preventDefault();
                    setPasswordDisabled(false);
                  }}
                />
              ) : (
                <input
                  type="submit"
                  className="profile-page--settings__user__form__input--submit"
                  value={t('SAVE_BTN') as string}
                  onClick={(e) => {
                    e.preventDefault();
                    setPasswordDisabled(true);
                    changePassword(newPassword);
                  }}
                />
              )}
            </form>
            <div className="profile-page--settings__btns">
              <button
                className="profile-page--settings__btns__btn--logout"
                onClick={() => {
                  signOut();
                }}
              >
                {t('LOG_OUT_BTN')}
              </button>
              <button
                className="profile-page--settings__btns__btn--delete"
                onClick={() => {
                  deleteUser();
                }}
              >
                {t('DELETE_ACCOUNT_BTN')}
              </button>
            </div>
          </section>
        </div>
      </main>
      <Footer />
    </>
  );
}
