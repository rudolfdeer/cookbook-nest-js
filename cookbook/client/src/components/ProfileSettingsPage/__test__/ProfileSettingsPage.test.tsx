import React from 'react';
import { shallow } from 'enzyme';
import ProfileSettingsPage from '../index';
import { ICookbook, IRecipe } from '../../../interfaces';

describe('profile settings page component', () => {
  const props = {
    user: {
      id: 1,
      name: 'Tet Test',
      image: 'images/user1.png',
      email: 'johndoe@test.com',
      password: 'user1',
      bio: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum',
      savedRecipes: [] as IRecipe[],
      savedCookbooks: [] as ICookbook[],
    },
    updateUser: jest.fn(),
    changeEmail: jest.fn(),
    changePassword: jest.fn(),
    signOut: jest.fn(),
    deleteUser: jest.fn(),
    updateUsersPhoto: jest.fn(),
  };

  const component = shallow(<ProfileSettingsPage {...props} />);

  it('renders the <ProfileSettingsPage /> component', () => {
    expect(component).toMatchSnapshot();
  });

  it('renders default image if no user photo passed', () => {
    const newProps = {
      ...props,
      user: {
        id: 1,
        name: 'Test Test',
        image: 'g',
        email: 'johndoe@test.com',
        password: 'user1',
        bio: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum',
        savedRecipes: [] as IRecipe[],
        savedCookbooks: [] as ICookbook[],
      },
    };
    const wrapper = shallow(<ProfileSettingsPage {...newProps} />);
    const img = wrapper.find('img');
    expect(img.prop('src')).toBe('./assets/images/photo-mask.png');
  });

  it('delete user on delete button click', () => {
    const btn = component.find('.profile-page--settings__btns__btn--delete');
    btn.simulate('click');
    expect(props.deleteUser).toHaveBeenCalled();
  });
});
