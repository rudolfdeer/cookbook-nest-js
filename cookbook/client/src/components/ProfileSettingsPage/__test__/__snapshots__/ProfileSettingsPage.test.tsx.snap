// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`profile settings page component renders the <ProfileSettingsPage /> component 1`] = `ShallowWrapper {}`;
